fix(iam): `Role.addManagedPolicy()` does not work for imported roles `IRole` #8307

stm29 opened the pull request:
### Issue #8307

Closes #8307

### Reason for this change
- This addManagedPolicy() just does nothing, neither throws warning nor fails.

### Description of changes
- This implements **addManagedPolicy()** for **ImportedRole**
- This throws Error, when `IRole` is used with `IManagedPolicy`, due to constrains

### Approaches decided
- we can't change **addManagedPolicy(policy: IManagedPolicy)** to **addManagedPolicy(policy: ManagedPolicy)** (i.e., to ManagedPolicy) , like how **attachInlinePolicy(policy: Policy)**, it will break for almost many customers.
  - https://github.com/aws/aws-cdk/blob/823ff6e03899f790a4cb1c43f92a02cc906ac356/packages/aws-cdk-lib/aws-iam/lib/identity-base.ts#L17-L21
  - So, we need to handle this gracefully when accepting only **ManagedPolicy**.
- We can't use IRole & IManagedPolicy to do the work at the same time, so we need to allow ManagedPolicy in params, and throw Error when IManagedPolicy is being sent, because of how **attachToRole()** will be used by **addManagedPolicy()**
- #### This PR Implements follows,
  - enables **addManagedPolicy()**
  - Accepts **ManagedPolicy**
  - Throws understandable Error, when `IRole` is used with `IManagedPolicy`.

### Description of how you validated changes
- unit & integration tests

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 4efe2f303ddc6e6199235d443c54e2b3da27c28a
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5b5ae80a-605e-40a2-8d13-c5d4f490a33f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

epoctic reviewed (approved):
Changes seem reasonable to me.

stm29 commented:
Can I get one more community review?

stm29 commented:
Communicated in #contributing channel. waiting for one more community review.

tmokmss commented on packages/aws-cdk-lib/aws-iam/lib/private/imported-role.ts:
```diff
@@ -74,8 +74,15 @@ export class ImportedRole extends Resource implements IRole, IComparablePrincipa
-    Annotations.of(this).addWarningV2('@aws-cdk/aws-iam:importedRoleManagedPolicyNotAdded', `Not adding managed policy: ${policy.managedPolicyArn} to imported role: ${this.roleName}`);
+  public addManagedPolicy(policy: ManagedPolicy): void {
+    try {
+      policy.attachToRole(this);
```
We should not change the signature (IManagedPolicy -> ManagedPolicy) because technically it is a breaking change.

Instead I'd write something like `(policy as ManagedPolicy).attachToRole`.

Or use type predicate:

```ts
    const isManagedPolicy = (x: IManagedPolicy): x is ManagedPolicy => {
      return (x as any).attachToRole !== undefined;
    };
    if (isManagedPolicy(policy)) {
      policy.attachToRole(this);
    }
```

This way we don't need catch the error for type check.

stm29 replied:
Resolved with **type predicate** and changed as **Warning** instead of **Error**.
Removed ```try ... catch```

tmokmss commented on packages/aws-cdk-lib/aws-s3/lib/notifications-resource/notifications-resource-handler.ts:
```diff
@@ -63,14 +63,20 @@ export class NotificationsResourceHandler extends Construct {
 
-    this.role.addManagedPolicy(
-      iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole'),
-    );
```
Why are we changing this file?

stm29 replied:
because `ManagedPolicy.addManagedPolicy` is being used in that file also, till now which is also **doing Nothing.**
```
      this.role.addManagedPolicy(
        iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole'),
      );
```

tmokmss replied:
I see. Maybe the maintainer will request you to move the fix to another PR because it is a related but different issue.

tmokmss replied:
Don't forget to revert this change :)

tmokmss commented on packages/aws-cdk-lib/aws-iam/lib/private/imported-role.ts:
```diff
@@ -74,8 +74,15 @@ export class ImportedRole extends Resource implements IRole, IComparablePrincipa
+      policy.attachToRole(this);
+    } catch (e) {
+      if (e instanceof Error && e.message === 'policy.attachToRole is not a function') {
+        throw new Error('Can\'t Combine IRole with IManagedPolicy. use ManagedPolicy directly.');
```
Throwing an error from this function is also a breaking change. Can we just add a warning as the previous behavior?

stm29 replied:
Done.

tmokmss reviewed (changes requested):
Thank you for the fix! I left some comments :)

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: f07668ad52e3337569d64118a33c837e235ca6fb
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6917f8e1-4350-41ec-b246-6adbdcbf07c5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: a83287358ecbdcd2b42e472a2531e0b262cf695b
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=319d8e5b-4e2e-467a-a544-d583ec7161de%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

tmokmss commented on packages/aws-cdk-lib/aws-s3/lib/notifications-resource/notifications-resource-handler.ts:
```diff
@@ -63,14 +63,20 @@ export class NotificationsResourceHandler extends Construct {
+    if (props.role) {
+      cdk.Annotations.of(this).addWarningV2('@aws-cdk-lib/aws-s3:missingLambdaExecutionRole', 'CDK doesn\'t add service-roles if `notificationsHandlerRole` is provided, Make Sure Imported Role have AWSLambdaBasicExecutionRole added');
+    }
+    if (!props.role) {
```
I'm not sure whether the conditions are correct. What happens when a user passed a Role construct? e.g. 
```ts
{
    role: new iam.Role(this, 'Role', {
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
    })
}
```

This should just add a managed policy to the role.

tmokmss replied:
Thanks for the clarification. Do we really need two warning for this? I feel one warning is enough. Then we can keep the current code as-is for this file.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 38f93e3d6c4391fa15a4b66195fbd2375ac737e5
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d36d7df7-6f9d-4140-8404-73c81c31c051%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

tmokmss commented on packages/aws-cdk-lib/aws-iam/lib/private/imported-role.ts:
```diff
@@ -75,7 +75,20 @@ export class ImportedRole extends Resource implements IRole, IComparablePrincipa
+    } else {
+      Annotations.of(this).addWarningV2(
+        '@aws-cdk/aws-iam:IRoleCantBeUsedWithIManagedPolicy',
+        'Can\'t combine IRole with IManagedPolicy. Use ManagedPolicy directly.',
```
We can add more information like `${policy.managedPolicyArn} to imported role: ${this.roleName}` as in the original code to make it clearer why and where the warning is emitted.

stm29 replied:
Corrected with One Common message.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 21fdf4a64f9172e3d873206d68419062f438ae11
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5c0844a1-405e-4db2-b1be-9bc7e6fef24b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 5c116fdcefe0d2ae467a60fa54807ed519d01273
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=eb2c8ee0-4a1f-489f-8fb1-5520c79d93ba%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

tmokmss reviewed (approved):
LGTM, thanks!

Leo10Gama commented:
Hi @stm29 thanks for your contribution! Reading through the changes, I'm trying to understand the nature of the changes being made. The PR description mentions you're adding functionality to `IRole`, but it looks like you're adding it to `ImportedRole`. Could you please update the description to better reflect this?

I also did some digging through the rest of the codebase, and it seems like some of the other extensions of `IRole` have other implementations of `addManagedPolicy()`, like [here](https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-iam/lib/private/precreated-role.ts#L110-L112):
```typescript
  public addManagedPolicy(policy: IManagedPolicy): void {
    this.managedPolicies.push(policy);
  }
```
Would a change like this work instead of the more complex logic?

It also looks like the changes in the S3 notifications resource are minor, so I would agree that moving that change to another PR and opening a separate issue would be best, as long as the tests that currently exist for it still pass with this change. Though I'm not quite sure what the issue is in this case. It seems like a few other libraries already call on the `addManagedPolicy()` method, so is there something specific to S3 in this regard?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 7513f825c788789ab9a2d41297a1ed27d03fe895
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d928454b-c86f-4aaf-a3ab-92397a7d4c68%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

stm29 commented:
@Leo10Gama , Thanks for the review.
>  I'm trying to understand the nature of the changes being made. The PR description mentions you're adding functionality to `IRole`, but it looks like you're adding it to `ImportedRole`. Could you please update the description to better reflect this?

Updated this with correct `ImportedRole`

----
> Would a change like this work instead of the more complex logic?

Indeed Logic is simple, following is the working Logic which adds Policy to the Role. and this is same as how other Implementations are already in place.
```
policy.attachToRole(this);
```
Other `if ... else` and `type predicate` is to make sure Implementation is working as expected.

-----
> It seems like a few other libraries already call on the addManagedPolicy() method, so is there something specific to S3 in this regard?

Sorry for the confusion, There is no change in `notifications-resource-handler.ts` file, I have reverted to `main`.

**Quick Question**: Should we need to add Unit Tests in all the possible occurrence of IManagedPolicy with IRole, or Just UnitTests in `packages/aws-cdk-lib/aws-iam/test/imported-role.test.ts` is sufficient?
If UnitTests in `imported-role.test.ts` is sufficient, then `notification.test.ts` can also be reverted to `main`

----

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: f5066afe8256d968bd81458f15dcf724d8f23715
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ca1af14c-939d-4266-8828-31e3f07b419b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 36deb38313017ca5b11dacb8d21e4c2c5b52e8e1
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f11cfa2e-d7a3-4c44-af04-4c215f3e2523%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
